File: js/src/convert.ts
/** Convert between protobuf messages and native TypeScript types. */

import type { Value } from "./generated/gql_types";
import type {
  GqlDate,
  GqlDuration,
  GqlEdge,
  GqlField,
  GqlLocalDateTime,
  GqlLocalTime,
  GqlNode,
  GqlPath,
  GqlRecord,
  GqlValue,
  GqlZonedDateTime,
  GqlZonedTime,
} from "./types";

/** Convert a protobuf Value to a native GqlValue. */
export function valueFromProto(v: Value): GqlValue {
  if (v.nullValue !== undefined) return null;
  if (v.booleanValue !== undefined) return v.booleanValue;
  if (v.integerValue !== undefined) return v.integerValue;
  if (v.unsignedIntegerValue !== undefined) return v.unsignedIntegerValue;
  if (v.floatValue !== undefined) return v.floatValue;
  if (v.stringValue !== undefined) return v.stringValue;
  if (v.bytesValue !== undefined) return v.bytesValue;

  if (v.dateValue) {
    return {
      year: v.dateValue.year,
      month: v.dateValue.month,
      day: v.dateValue.day,
    } satisfies GqlDate;
  }

  if (v.localTimeValue) {
    return {
      hour: v.localTimeValue.hour,
      minute: v.localTimeValue.minute,
      second: v.localTimeValue.second,
      nanosecond: v.localTimeValue.nanosecond,
    } satisfies GqlLocalTime;
  }

  if (v.zonedTimeValue) {
    const t = v.zonedTimeValue.time!;
    return {
      time: {
        hour: t.hour,
        minute: t.minute,
        second: t.second,
        nanosecond: t.nanosecond,
      },
      offsetMinutes: v.zonedTimeValue.offsetMinutes,
    } satisfies GqlZonedTime;
  }

  if (v.localDatetimeValue) {
    const d = v.localDatetimeValue.date!;
    const t = v.localDatetimeValue.time!;
    return {
      date: { year: d.year, month: d.month, day: d.day },
      time: {
        hour: t.hour,
        minute: t.minute,
        second: t.second,
        nanosecond: t.nanosecond,
      },
    } satisfies GqlLocalDateTime;
  }

  if (v.zonedDatetimeValue) {
    const d = v.zonedDatetimeValue.date!;
    const t = v.zonedDatetimeValue.time!;
    return {
      date: { year: d.year, month: d.month, day: d.day },
      time: {
        hour: t.hour,
        minute: t.minute,
        second: t.second,
        nanosecond: t.nanosecond,
      },
      offsetMinutes: v.zonedDatetimeValue.offsetMinutes,
    } satisfies GqlZonedDateTime;
  }

  if (v.durationValue) {
    return {
      months: v.durationValue.months,
      nanoseconds: v.durationValue.nanoseconds,
    } satisfies GqlDuration;
  }

  if (v.listValue) {
    return v.listValue.elements.map(valueFromProto);
  }

  if (v.recordValue) {
    const fields: GqlField[] = v.recordValue.fields.map((f) => ({
      name: f.name,
      value: f.value ? valueFromProto(f.value) : null,
    }));
    return { fields } satisfies GqlRecord;
  }

  if (v.nodeValue) {
    const props = new Map<string, GqlValue>();
    for (const [k, pv] of Object.entries(v.nodeValue.properties)) {
      props.set(k, valueFromProto(pv));
    }
    return {
      id: v.nodeValue.id,
      labels: v.nodeValue.labels,
      properties: props,
    } satisfies GqlNode;
  }

  if (v.edgeValue) {
    const props = new Map<string, GqlValue>();
    for (const [k, pv] of Object.entries(v.edgeValue.properties)) {
      props.set(k, valueFromProto(pv));
    }
    return {
      id: v.edgeValue.id,
      labels: v.edgeValue.labels,
      sourceNodeId: v.edgeValue.sourceNodeId,
      targetNodeId: v.edgeValue.targetNodeId,
      undirected: v.edgeValue.undirected,
      properties: props,
    } satisfies GqlEdge;
  }

  if (v.pathValue) {
    const nodes: GqlNode[] = v.pathValue.nodes.map((n) => {
      const props = new Map<string, GqlValue>();
      for (const [k, pv] of Object.entries(n.properties)) {
        props.set(k, valueFromProto(pv));
      }
      return { id: n.id, labels: n.labels, properties: props };
    });
    const edges: GqlEdge[] = v.pathValue.edges.map((e) => {
      const props = new Map<string, GqlValue>();
      for (const [k, pv] of Object.entries(e.properties)) {
        props.set(k, valueFromProto(pv));
      }
      return {
        id: e.id,
        labels: e.labels,
        sourceNodeId: e.sourceNodeId,
        targetNodeId: e.targetNodeId,
        undirected: e.undirected,
        properties: props,
      };
    });
    return { nodes, edges } satisfies GqlPath;
  }

  // BigInteger, BigFloat, Decimal - not supported in v0.1
  return null;
}

/** Convert a native value to a protobuf Value. */
export function valueToProto(value: GqlValue): Value {
  if (value === null) return { nullValue: {} };
  if (typeof value === "boolean") return { booleanValue: value };
  if (typeof value === "bigint") return { integerValue: value };
  if (typeof value === "number") return { floatValue: value };
  if (typeof value === "string") return { stringValue: value };
  if (value instanceof Uint8Array) return { bytesValue: value };
  if (Array.isArray(value)) {
    return { listValue: { elements: value.map(valueToProto) } };
  }
  return { nullValue: {} };
}

File: go/convert.go
package gwp

import (
	pb "github.com/GrafeoDB/gql-wire-protocol/go/gen/gql"
)

// valueFromProto converts a protobuf Value to a native Go value.
func valueFromProto(v *pb.Value) any {
	if v == nil {
		return nil
	}

	switch k := v.Kind.(type) {
	case *pb.Value_NullValue:
		return nil
	case *pb.Value_BooleanValue:
		return k.BooleanValue
	case *pb.Value_IntegerValue:
		return k.IntegerValue
	case *pb.Value_UnsignedIntegerValue:
		return k.UnsignedIntegerValue
	case *pb.Value_FloatValue:
		return k.FloatValue
	case *pb.Value_StringValue:
		return k.StringValue
	case *pb.Value_BytesValue:
		return k.BytesValue
	case *pb.Value_DateValue:
		d := k.DateValue
		return &GqlDate{Year: d.Year, Month: d.Month, Day: d.Day}
	case *pb.Value_LocalTimeValue:
		t := k.LocalTimeValue
		return &GqlLocalTime{
			Hour: t.Hour, Minute: t.Minute,
			Second: t.Second, Nanosecond: t.Nanosecond,
		}
	case *pb.Value_ZonedTimeValue:
		zt := k.ZonedTimeValue
		t := zt.Time
		return &GqlZonedTime{
			Time: GqlLocalTime{
				Hour: t.Hour, Minute: t.Minute,
				Second: t.Second, Nanosecond: t.Nanosecond,
			},
			OffsetMinutes: zt.OffsetMinutes,
		}
	case *pb.Value_LocalDatetimeValue:
		ldt := k.LocalDatetimeValue
		d := ldt.Date
		t := ldt.Time
		return &GqlLocalDateTime{
			Date: GqlDate{Year: d.Year, Month: d.Month, Day: d.Day},
			Time: GqlLocalTime{
				Hour: t.Hour, Minute: t.Minute,
				Second: t.Second, Nanosecond: t.Nanosecond,
			},
		}
	case *pb.Value_ZonedDatetimeValue:
		zdt := k.ZonedDatetimeValue
		d := zdt.Date
		t := zdt.Time
		return &GqlZonedDateTime{
			Date: GqlDate{Year: d.Year, Month: d.Month, Day: d.Day},
			Time: GqlLocalTime{
				Hour: t.Hour, Minute: t.Minute,
				Second: t.Second, Nanosecond: t.Nanosecond,
			},
			OffsetMinutes: zdt.OffsetMinutes,
		}
	case *pb.Value_DurationValue:
		dur := k.DurationValue
		return &GqlDuration{Months: dur.Months, Nanoseconds: dur.Nanoseconds}
	case *pb.Value_ListValue:
		elems := make([]any, len(k.ListValue.Elements))
		for i, e := range k.ListValue.Elements {
			elems[i] = valueFromProto(e)
		}
		return elems
	case *pb.Value_RecordValue:
		fields := make([]GqlField, len(k.RecordValue.Fields))
		for i, f := range k.RecordValue.Fields {
			fields[i] = GqlField{Name: f.Name, Value: valueFromProto(f.Value)}
		}
		return &GqlRecord{Fields: fields}
	case *pb.Value_NodeValue:
		n := k.NodeValue
		props := make(map[string]any, len(n.Properties))
		for key, pv := range n.Properties {
			props[key] = valueFromProto(pv)
		}
		return &GqlNode{ID: n.Id, Labels: n.Labels, Properties: props}
	case *pb.Value_EdgeValue:
		e := k.EdgeValue
		props := make(map[string]any, len(e.Properties))
		for key, pv := range e.Properties {
			props[key] = valueFromProto(pv)
		}
		return &GqlEdge{
			ID: e.Id, Labels: e.Labels,
			SourceNodeID: e.SourceNodeId, TargetNodeID: e.TargetNodeId,
			Undirected: e.Undirected, Properties: props,
		}
	case *pb.Value_PathValue:
		p := k.PathValue
		nodes := make([]*GqlNode, len(p.Nodes))
		for i, n := range p.Nodes {
			props := make(map[string]any, len(n.Properties))
			for key, pv := range n.Properties {
				props[key] = valueFromProto(pv)
			}
			nodes[i] = &GqlNode{ID: n.Id, Labels: n.Labels, Properties: props}
		}
		edges := make([]*GqlEdge, len(p.Edges))
		for i, e := range p.Edges {
			props := make(map[string]any, len(e.Properties))
			for key, pv := range e.Properties {
				props[key] = valueFromProto(pv)
			}
			edges[i] = &GqlEdge{
				ID: e.Id, Labels: e.Labels,
				SourceNodeID: e.SourceNodeId, TargetNodeID: e.TargetNodeId,
				Undirected: e.Undirected, Properties: props,
			}
		}
		return &GqlPath{Nodes: nodes, Edges: edges}
	default:
		return nil
	}
}

// valueToProto converts a native Go value to a protobuf Value.
func valueToProto(value any) *pb.Value {
	if value == nil {
		return &pb.Value{Kind: &pb.Value_NullValue{NullValue: &pb.NullValue{}}}
	}

	switch v := value.(type) {
	case bool:
		return &pb.Value{Kind: &pb.Value_BooleanValue{BooleanValue: v}}
	case int64:
		return &pb.Value{Kind: &pb.Value_IntegerValue{IntegerValue: v}}
	case int:
		return &pb.Value{Kind: &pb.Value_IntegerValue{IntegerValue: int64(v)}}
	case float64:
		return &pb.Value{Kind: &pb.Value_FloatValue{FloatValue: v}}
	case string:
		return &pb.Value{Kind: &pb.Value_StringValue{StringValue: v}}
	case []byte:
		return &pb.Value{Kind: &pb.Value_BytesValue{BytesValue: v}}
	case []any:
		elems := make([]*pb.Value, len(v))
		for i, e := range v {
			elems[i] = valueToProto(e)
		}
		return &pb.Value{Kind: &pb.Value_ListValue{ListValue: &pb.GqlList{Elements: elems}}}
	default:
		return &pb.Value{Kind: &pb.Value_NullValue{NullValue: &pb.NullValue{}}}
	}
}
